consensus/istanbul/uptime/uptime.go

```diff
@@ -0,0 +1,242 @@
+	// We can't include that epoch's last block as part of the tally because the epoch val score is calculated
+	// using a tally that is updated AFTER a block is finalized.
+	// Note that it's possible to count up to the last block of the epoch, but it's much harder to implement
+	// than couting up to the second to last one.
+	lastBlockToMonitor := epochLastBlock - 1
+
+	return firstBlockToMonitor, lastBlockToMonitor
+}
+
+// Monitor is responsible for monitoring uptime
+// by processing blocks
+type Monitor struct {
+	epochSize      uint64
+	lookbackWindow uint64
+
+	logger log.Logger
+	store  Store
+}
+
+// NewMonitor creates a new uptime monitor
+func NewMonitor(store Store, epochSize, lookbackWindow uint64) *Monitor {
+	return &Monitor{
+		epochSize:      epochSize,
+		lookbackWindow: lookbackWindow,
+		store:          store,
+		logger:         log.New("module", "uptime-monitor"),
+	}
+}
+
+// MonitoringWindow returns the monitoring window for the given epoch in the format
+// [firstBlock, lastBlock] both inclusive
+func (um *Monitor) MonitoringWindow(epoch uint64) (uint64, uint64) {
+	return GetUptimeMonitoringWindow(epoch, um.epochSize, um.lookbackWindow)
+}
+
+// MonitoringWindowSize returns the size (in blocks) of the monitoring window
+func (um *Monitor) MonitoringWindowSize(epoch uint64) uint64 {
+	start, end := um.MonitoringWindow(epoch)
+	return end - start + 1
+}
+
+// ComputeValidatorsUptime retrieves the uptime score for each validator for a given epoch
+func (um *Monitor) ComputeValidatorsUptime(epoch uint64, valSetSize int) ([]*big.Int, error) {
+	logger := um.logger.New("func", "Backend.updateValidatorScores", "epoch", epoch)
+	logger.Trace("Updating validator scores")
+
+	// The denominator is the (last block - first block + 1) of the val score tally window
+	denominator := um.MonitoringWindowSize(epoch)
+
+	uptimes := make([]*big.Int, 0, valSetSize)
+	accumulated := um.store.ReadAccumulatedEpochUptime(epoch)
+
+	if accumulated == nil {
+		err := errors.New("Accumulated uptimes not found, cannot update validator scores")
+		logger.Error(err.Error())
+		return nil, err
+	}
+
+	for i, entry := range accumulated.Entries {
+		if i >= valSetSize {
+			break
+		}
+
+		if entry.ScoreTally > denominator {
+			logger.Error("ScoreTally exceeds max possible", "scoreTally", entry.ScoreTally, "denominator", denominator, "valIdx", i)
+			uptimes = append(uptimes, params.Fixidity1)
+			continue
+		}
+
+		numerator := big.NewInt(0).Mul(big.NewInt(int64(entry.ScoreTally)), params.Fixidity1)
+		uptimes = append(uptimes, big.NewInt(0).Div(numerator, big.NewInt(int64(denominator))))
+	}
+
+	if len(uptimes) < valSetSize {
+		err := fmt.Errorf("%d accumulated uptimes found, cannot update validator scores", len(uptimes))
+		logger.Error(err.Error())
+		return nil, err
+	}
+
+	return uptimes, nil
+}
+
+// ProcessBlock will monitor uptimeness information on the give block and add it to current epoch tally
```

Comment:
I'd rephrase this comment as:
```
// ProcessBlock uses the block's signature bitmap (which encodes who signed the parent block) to update the epoch's Uptime data
```

If I'm confused about the bitmap being about the parent block, then modify accordingly.